c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -209,6 +254,32 @@ class DeviceCachingAllocator {
 
   bool set_fraction = false;
 
+  // Members specific to CUDA graphs
+
+  struct PrivatePool {
+    PrivatePool() : use_count(1),
+                    large_blocks(BlockComparator),
+                    small_blocks(BlockComparator) {}
+    // Number of live graphs using this pool
+    int use_count;
+    // Number of unfreed cudaMallocs made for this pool. When use_count and cudaMalloc_count
+    // drop to zero, we can delete this PrivatePool from graph_pools.
+    int cudaMalloc_count;
+    // Instead of maintaining private BlockPools here, I could stuff all blocks (private or no)
+    // into the top-level large_blocks and small_blocks, and distinguish private blocks by adding
+    // a "pool id" check above the stream check in BlockComparator. BlockComparator is performance-
+    // critial though, I'd rather not add more logic to it.
+    BlockPool large_blocks;
+    BlockPool small_blocks;
+  };
+
+  // Private pools for CUDA graphs
+  std::unordered_map<CUDACaptureid_t/*pool id*/, PrivatePool> graph_pools;
+
+  // Maps a capturing stream to its assigned private pool,
+  // in case we want multiple captures to share the same pool
+  std::unordered_map<CUDACaptureid_t/*capture id*/, CUDACaptureid_t/*pool id*/> capture_to_pool_map;
```

Comment:
You sure you want to use an `unordered_map` here? They're famously slow. I guess this only slows you down when capturing so it's nbd.

Reply:
Capture ids are assigned to each capture by Cuda, we can't predict their values (In Cuda today, I think they happen to be sequential starting from 1, but that's not guaranteed and may change). So afaict we need unordered_map, but im open to better ideas. As you say it only affects capture so performance isn't crucial.

Reply:
if you need a non-crappy hash map `flat_hash_map` may be of use

Reply:
For this case i think unordered map is alright, but I don't mind changing to `flat_hash_map` if you have strong feelings about it.